rust/lance-linalg/src/simd/u8.rs

```diff
@@ -0,0 +1,372 @@
+    }
+}
+
+impl<'a> From<&'a [u8; 16]> for u8x16 {
+    fn from(value: &'a [u8; 16]) -> Self {
+        unsafe { Self::load_unaligned(value.as_ptr()) }
+    }
+}
+
+impl SIMD<u8, 16> for u8x16 {
+    #[inline]
+
+    fn splat(val: u8) -> Self {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            Self(_mm_set1_epi8(val as i8))
+        }
+        #[cfg(target_arch = "aarch64")]
+        unsafe {
+            Self(vdupq_n_u8(val))
+        }
+    }
+
+    #[inline]
+    fn zeros() -> Self {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            Self(_mm_setzero_si128())
+        }
+        #[cfg(target_arch = "aarch64")]
+        {
+            Self::splat(0)
+        }
+    }
+
+    #[inline]
+    unsafe fn load(ptr: *const u8) -> Self {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            Self(_mm_loadu_si128(ptr as *const __m128i))
+        }
+        #[cfg(target_arch = "aarch64")]
+        {
+            Self::load_unaligned(ptr)
+        }
+    }
+
+    #[inline]
+    unsafe fn load_unaligned(ptr: *const u8) -> Self {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            Self(_mm_loadu_si128(ptr as *const __m128i))
+        }
+        #[cfg(target_arch = "aarch64")]
+        {
+            Self(vld1q_u8(ptr))
+        }
+    }
+
+    #[inline]
+    unsafe fn store(&self, ptr: *mut u8) {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            _mm_storeu_si128(ptr as *mut __m128i, self.0)
+        }
+        #[cfg(target_arch = "aarch64")]
+        unsafe {
+            vst1q_u8(ptr, self.0)
+        }
+    }
+
+    #[inline]
+
+    unsafe fn store_unaligned(&self, ptr: *mut u8) {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            _mm_storeu_si128(ptr as *mut __m128i, self.0)
+        }
+        #[cfg(target_arch = "aarch64")]
+        unsafe {
+            vst1q_u8(ptr, self.0)
+        }
+    }
+
+    fn reduce_sum(&self) -> u8 {
+        todo!("the signature of reduce_sum is not correct");
+        // #[cfg(target_arch = "x86_64")]
+        // unsafe {
+        //     let zeros = _mm_setzero_si128();
+        //     let sum = _mm_sad_epu8(self.0, zeros);
+
+        //     let lower = _mm_cvtsi128_si64(sum) as u32;
+        //     let upper = _mm_extract_epi64(sum, 1) as u32;
+        //     lower + upper
+        // }
+        // #[cfg(target_arch = "aarch64")]
+        // unsafe {
+        //     let low = vget_low_u8(self.0);
+        //     let high = vget_high_u8(self.0);
+        //     let sum = vaddl_u8(low, high);
+        //     let sum16 = vaddw_u16(vdupq_n_u32(0), sum);
+        //     let total = vpadd_u32(vget_low_u32(sum16), vget_high_u32(sum16));
+        //     vget_lane_u32(total, 0) + vget_lane_u32(total, 1)
+        // }
+    }
+
+    #[inline]
+    fn reduce_min(&self) -> u8 {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            let low = _mm_and_si128(self.0, _mm_set1_epi8(0xFF_u8 as i8));
```

Comment:
this is only using sse1?  Curious whether there are `avx2` related coding to make this even faster.

Reply:
didn't find a `avx2` intrinsic to do this, but `reduce_min` is not used for now

Reply:
Lets just delete `reduce_sum` and `reduce_min` if they are not used.